### PR TITLE
fix: hamt directory listings

### LINKED ISSUE #192: Gateway fails to load ipfs wiki homepage
STR:
- Load http://localhost:9050/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq/wiki/

Expected:
- Get wikipedia homepage.

Observed:
- 500 error, link 'wiki' not found from the UnixfsNode::HamtShard

### PR DESCRIPTION
**Pull request by dignifiedquire**
- Actually load and list out all children of a hamt shard when viewing it on the gateway
- Split out directory listings and file content rendering
- Streaming loading of hamt directory listings
- Use correct sizes for unixfs files, and provide sizes on multipart files

Closes #192

### CODE REVIEW

File: iroh-gateway/src/core.rs [hunk 1]
```diff
@@ -474,82 +479,87 @@ async fn serve_fs(
...
+            let dir_list: anyhow::Result<Vec<_>> = res
+                .unixfs_read_dir(
+                    &state.client.resolver,
+                    OutMetrics {
+                        metrics: state.metrics.clone(),
+                        start: start_time,
+                    },
+                )
+                .map_err(|e| error(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string(), &state))?
+                .expect("already known this is a directory")
```
Comment (Arqu) [inline iroh-gateway/src/core.rs:500]: I don't really understand what this `expect` says.

File: iroh-resolver/src/resolver.rs [hunk 2]
```diff
@@ -269,18 +285,38 @@ pub enum UnixfsType {
...
 
+impl<T: ContentLoader> OutPrettyReader<T> {
+    /// Returrns the size in bytes, if known in advance.
+    pub fn size(&self) -> Option<u64> {
+        match self {
+            OutPrettyReader::DagPb(reader)
+            | OutPrettyReader::DagCbor(reader)
+            | OutPrettyReader::DagJson(reader)
+            | OutPrettyReader::Raw(reader) => reader.size(),
+            OutPrettyReader::Unixfs(reader) => reader.size(),
```
Comment (Arqu) [inline iroh-resolver/src/resolver.rs:302]: Why is this line separate?

**Review by Arqu — APPROVED** [review]
Not going to pretend I understood it all, but codewise looks good and logic seems fine. If it does what it says on the tin then 👍

**Comment by Arqu** [comment]
Also CI is pretty sad

File: iroh-resolver/src/resolver.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs:302]: because it is a different type

File: iroh-gateway/src/core.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-gateway/src/core.rs:500]: this unwraps the option, which is guranteed to be `Some` if it is a directory like structure

**Review by Arqu — DISMISSED** [review]

File: iroh-gateway/src/core.rs (hunk 1 above)
Comment (ramfox) [reply iroh-gateway/src/core.rs:500]: maybe change to "This will always unwrap since we have already guaranteed it to be a directory"

or "Should never panic since it is guaranteed to be a directory"

Cause I agree, the phrasing is hard to parse

File: iroh-resolver/src/resolver.rs [hunk 3]
```diff
@@ -269,18 +285,38 @@ pub enum UnixfsType {
...
     DagPb(BytesReader),
-    Unixfs(UnixfsReader<T>),
+    Unixfs(UnixfsContentReader<T>),
     DagCbor(BytesReader),
     DagJson(BytesReader),
     Raw(BytesReader),
 }
 
+impl<T: ContentLoader> OutPrettyReader<T> {
+    /// Returrns the size in bytes, if known in advance.
```
Comment (ramfox) [inline iroh-resolver/src/resolver.rs:295]: "Returrns" => "Returns"

File: iroh-resolver/src/hamt.rs [hunk 4]
```diff
@@ -68,21 +87,66 @@ impl InnerNode {
...
+                        }
+                    }
+                    UnixfsNode::HamtShard(_, hamt) => {
+                        let mut children = hamt.children(loader);
+                        while let Some(link) = children.next().await {
+                            let link = link?;
+                            yield link;
+                        }
+                    }
+                    _ => {}
```
Comment (ramfox) [inline iroh-resolver/src/hamt.rs:145]: why is it expected that only the `UnixfsNode::Directory` and `UnixfsNode::HamtShard` will have children? Isn't it possible for `UnixfsNode::File`s to also have links?

I'm probably missing a subtly here.

**Review by ramfox — CHANGES_REQUESTED** [review]
A few typos on comments and some questions!

File: iroh-resolver/src/unixfs.rs [hunk 5]
```diff
@@ -226,23 +249,40 @@ impl UnixfsNode {
...
             | UnixfsNode::File(node)
             | UnixfsNode::Symlink(node)
-            | UnixfsNode::HamtShard(node) => node.size(),
+            | UnixfsNode::HamtShard(node, _) => node.size(),
         }
     }
 
-    pub fn links(&self) -> Links {
+    /// Returns the filesize in bytes.
+    /// Should only be set for `Raw` and `File`.
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs.rs:257]: it fully proxies the underlying value atm, but the expectation is that that is not set in the protobuf for non file types

File: iroh-resolver/src/hamt.rs (hunk 4 above)
Comment (dignifiedquire) [reply iroh-resolver/src/hamt.rs:145]: The subtly here is my bad naming skillz. "children" here means things in a directory, not children as in links

File: iroh-resolver/src/resolver.rs [hunk 6]
```diff
@@ -269,18 +285,38 @@ pub enum UnixfsType {
...
+}
+
 pub struct BytesReader {
     pos: usize,
     bytes: Bytes,
     om: OutMetrics,
 }
 
+impl BytesReader {
+    /// Returrns the size in bytes, if known in advance.
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs]: fixed

File: iroh-resolver/src/resolver.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs]: fixed
